### PR TITLE
Improve handling of changed alloc config.

### PR DESCRIPTION
**Pull request by toregge**
@baldersheim : please review

### CODE REVIEW

File: searchcore/src/tests/proton/common/alloc_config/alloc_config_test.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    return GrowStrategy(initial_docs, 0.1, 1, 0.15);
+}
+
+AllocStrategy make_alloc_strategy(uint32_t initial_docs) {
+    return AllocStrategy(make_grow_strategy(initial_docs), baseline_compaction_strategy, 10000);
+}
+
+};
+
+TEST(AllocConfigTest, hello_world)
```
Comment (geirst) [inline searchcore/src/tests/proton/common/alloc_config/alloc_config_test.cpp:27]: Consider a more descriptive test name.

File: searchcore/src/vespa/searchcore/proton/common/alloc_config.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "alloc_strategy.h"
+
+namespace proton {
+
+enum class SubDbType;
+
+/*
+ * Class representing allocation config for proton which can be used
+ * to make an allocation strategy.
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/common/alloc_config.h:13]: Consider adding some details on what kind of allocations we are talking about.

File: searchcore/src/vespa/searchcore/proton/common/alloc_strategy.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#pragma once
+
+#include <vespa/searchcommon/common/compaction_strategy.h>
+#include <vespa/searchcommon/common/growstrategy.h>
+#include <iosfwd>
+
+namespace proton {
+
+/*
+ * Class representing allocation strategy
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/common/alloc_strategy.h:12]: Consider adding some details on what kind of allocations we are talking about.

File: searchcore/src/vespa/searchcore/proton/server/documentdbconfigmanager.cpp [hunk 4]
```diff
@@ -382,6 +398,10 @@ DocumentDBConfigManager::update(const ConfigSnapshot &snapshot)
         (*new_threading_service_config == *old_threading_service_config)) {
         new_threading_service_config = old_threading_service_config;
     }
+    auto new_alloc_config = build_alloc_config(_bootstrapConfig->getProtonConfig(), _docTypeName);
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentdbconfigmanager.cpp:401]: Consider if this change should be tested in src/tests/proton/proton_config_fetcher/proton_config_fetcher_test.cpp.

File: searchcore/src/vespa/searchcore/proton/server/reconfig_params.cpp [hunk 5]
```diff
@@ -51,7 +52,7 @@ ReconfigParams::shouldIndexManagerChange() const
 bool
 ReconfigParams::shouldAttributeManagerChange() const
 {
-    return _res.attributesChanged || _res.importedFieldsChanged || _res.visibilityDelayChanged;
+    return _res.attributesChanged || _res.importedFieldsChanged || _res.visibilityDelayChanged || _res.alloc_config_changed;
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/reconfig_params.cpp:55]: Should probably be tested in searchcore/src/tests/proton/documentdb/configurer/configurer_test.cpp.

File: searchlib/src/vespa/searchlib/attribute/attributevector.cpp [hunk 6]
```diff
@@ -769,6 +770,31 @@ AttributeVector::logEnumStoreEvent(const char *reason, const char *stage)
...
+        updateStat(true);
+        if (_status.getOnHold() <= hold_limit) {
+            return;
+        }
+        std::this_thread::sleep_for(retry < 20 ? 20ms : 100ms);
+    }
+}
+
+void
+AttributeVector::update_config(const Config& cfg)
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/attribute/attributevector.cpp:788]: We should add a unit test for this on the attribute vector level.

**Review by geirst — COMMENTED** [review]
Very good. Some comments to look at.

File: searchcore/src/tests/proton/common/alloc_config/alloc_config_test.cpp (hunk 1 above)
Comment (toregge) [reply searchcore/src/tests/proton/common/alloc_config/alloc_config_test.cpp]: Fixed.

File: searchcore/src/vespa/searchcore/proton/common/alloc_config.h (hunk 2 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/common/alloc_config.h]: Class comment has been adjusted.

File: searchcore/src/vespa/searchcore/proton/common/alloc_strategy.h (hunk 3 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/common/alloc_strategy.h]: Class comment has been adjusted.

File: searchlib/src/vespa/searchlib/attribute/attributevector.cpp (hunk 6 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/attribute/attributevector.cpp:788]: It is unit tested, c.f. attribute_compaction_test.cpp.

File: searchcore/src/vespa/searchcore/proton/server/reconfig_params.cpp (hunk 5 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/server/reconfig_params.cpp:55]: Done.

**Review by baldersheim — APPROVED** [review]

**Comment by baldersheim** [comment]
Very good, merge at will

**Pull request merged**